datajunction-server/datajunction_server/construction/build_v2.py

```diff
@@ -179,202 +184,398 @@ async def get_measures_query(  # pylint: disable=too-many-locals
+        )
+        if physical_table and not self._filters and not self.dimensions:
+            physical_table.parenthesized = False
+            return ast.Query(
+                select=ast.Select(
+                    projection=physical_table.columns,  # type: ignore
+                    from_=ast.From(relations=[ast.Relation(physical_table)]),
+                ),
+            )
 
-            dimension_node_query = await build_dimension_node_query(
-                session,
-                build_criteria,
-                link,
-                to_filter_asts(filters),
-                cte_mapping,
+        node_ast = (
+            await compile_node_ast(self.session, self.node_revision)
+            if not physical_table
+            else ast.Query(
+                select=ast.Select(
+                    projection=physical_table.columns,  # type: ignore
+                    from_=ast.From(relations=[ast.Relation(physical_table)]),
+                ),
             )
-            dimension_join.node_query = convert_to_cte(
-                dimension_node_query,
-                final_ast,
-                link.dimension.name,
+        )
+
+        # Find all dimension node joins necessary for the requested dimensions and filters
+        dimension_node_joins = await self.find_dimension_node_joins()
+
+        ###################
+        ## Build Node CTEs
+        ###################
+
+        # Start tracking node query CTEs as they get built
+        cte_mapping: Dict[str, ast.Query] = {}  # Maps node name to its CTE
+
+        # Build node AST into a CTE
+        ctx = CompileContext(self.session, DJException())
+        await node_ast.compile(ctx)
+        self.errors.extend(ctx.exception.errors)
+
+        node_alias = ast.Name(amenable_name(self.node_revision.name))
+        node_ast = await build_ast(
+            self.session,
+            self.node_revision,
+            node_ast,
+            filters=self._filters,
+            build_criteria=self._build_criteria,
+            ctes_mapping=cte_mapping,
+        )
+
+        # Initialize the final query AST structure
+        final_ast = ast.Query(
+            select=ast.Select(
+                projection=[
+                    ast.Column(
+                        ast.Name(col.alias_or_name.name),  # type: ignore
+                        _table=node_ast,
+                        _type=col.type,  # type: ignore
+                    )
+                    for col in node_ast.select.projection
+                ],
+                from_=ast.From(relations=[ast.Relation(node_alias)]),  # type: ignore
+            ),
+            ctes=[*node_ast.ctes, node_ast],
+        )
+        node_ast.ctes = []
+        node_ast = node_ast.to_cte(node_alias)
+        cte_mapping[self.node_revision.name] = node_ast
+
+        # Start building the dimension joins and adding them to the CTEs
+        for _, dimension_join in dimension_node_joins.items():
```

Comment:
This feels like it would go nicely in a build step method on this class. It looks like it's the only place `dimension_node_joins` defined above is used so you can even wrap line 399 into this. Then here it would just be a single self-documenting line along the lines of `self.add_dimensions_as_ctes()`

Reply:
@samredai I'd originally avoided the self-documenting submethods because I was trying to avoid the pattern of modifying arguments that are passed into functions. However, I do agree that it'll be easier to read, so what I ended up doing was creating a set of attributes on the `QueryBuilder` that will be modified as the query gets built (so that it's not a surprise that some arguments will be changed), and then just return the final query at the end of `build()`.

Reply:
Makes sense! Thanks @shangyian